### ISSUE TITLE
Some questions to discuss about `IdxDomain` in Base Analysis

### ISSUE THREAD

**Issue by michael-schwarz** [goblint/analyzer#241]
While trying to insert bounds checking for arrays, @vandah @Wherekonshade @EdinCitaku stumbled across the issue that even in code such as 

```
int arr[5] = {0,0,0,0,0};
for(int i=0;i < 3;i++) {
    arr[i] = 5;
}
```

the parameter passed to `CArrays.get` for the index from the IdxDomain is `T` (even in the final iteration after narrowing terminated).

The cause is this function
https://github.com/goblint/analyzer/blob/c989e10cca8d6e7c6f451ee9d7394b0ccaddb9d0/src/analyses/base.ml#L95-L98

that is called from

https://github.com/goblint/analyzer/blob/c989e10cca8d6e7c6f451ee9d7394b0ccaddb9d0/src/analyses/base.ml#L733-L744

during `eval_rv` and `eval_lv`.

**This amounts to having only definite offsets or top offsets in `IdxDomain`**. This raises several questions:

- **Why do we not want any Intervals or whatever the enabled IntDomains can handle in the IdxDomain?** I assume this is for performance reasons to keep the number of possible addresses small? If so, **is this still required after switching to a HoareDomain for addresses?** It would seem that this alone might address the address issue (pun intended) with too many different offsets? How can one end up with a high number of different intervals (of size bigger than 1) where there is no subsumption?
- If we do not want Intervals in the IdxDomain, we should probably switch the IdxDomain to be a domain that properly enforces it instead of doing it in this wishy-washy way?
- Should maybe the IntDomains that are used in the IdxDomain be configurable on their own as a separate parameter for tuning precision vs cost? Then one could e.g. enable intervals for offsets if one wants to check out-of-bounds accesses and disable them by default?

**Re Bounds-Checking:** The `get` and `set` functions for `CArrays` already also get an `exp option` for the index expression, so in many cases (but not all as it is `None` sometimes) the immediate fix is issuing a query to evaluate this expression, obtain some interval that is more precise and then using that to do bounds-checking.

Nevertheless, I guess the overall issue is still worth discussing.

**Comment by vogler** [comment]
(1) Probably didn't notice this after changing to HoareDomain. Should be fine to allow any IntDomain now.
(3) We could add some option for it. Not sure if it would easily work though - if you use a variable to index and don't have an interval for it, where should it come from for the index expression?
If it works out anyway, the default should be 'default' to use the same int domains as are activated for base.

**Comment by michael-schwarz** [comment]
> Probably didn't notice this after changing to HoareDomain. Should be fine to allow any IntDomain now.

I'll update this and run some benchmarks to see what effect it has.

> We could add some option for it. Not sure if it would easily work though - if you use a variable to index and don't have an interval for it, where should it come from for the index expression?

The usual case would be having a subset of the normal int domains for addresses. In the other case, it would often end up being top. However, a case where there might be some difference is if one has e.g. `arr[0]` and `arr[1]` in the address domain. Without intervals that would be `arr[T]`, with intervals it could be `arr[0,1]`. Maybe something to keep in mind but implementing it is not the most pressing thing.

**Comment by vogler** [comment]
Yea, some thoughts:
1. fewer int domains for addresses trade precision for speed-up,
2. more int domains than base only makes a difference for constants. Something like `a[i]` and `a[i+1]` would not benefit from intervals in offsets only.

(2) probably doesn't make much of a difference.
(1) probably doesn't speed up much since addresses are not modified often compared to e.g. ints, and we usually want addresses to be as precise as possible, so one would first try other options.

**Comment by michael-schwarz** [comment]
I did a testrun of sv-comp, it did not really make a meaningful difference either way.
Went with allowing the full ID in IdxDomain:

~~~OCAML
let iDtoIdx = ID.cast_to (Cilfacade.ptrdiff_ikind ())
~~~

Closed via 542a600778fce059bec649506f513b5a8a7ef651
